docs/proposals/202004_embedd_cortex_frontend.md

```diff
@@ -0,0 +1,128 @@
+
+* `--query-range.split-interval`, `time.Duration`
+* `--query-range.max-retries-per-request`, `int`, default = `5`
+* `--query-range.disable-step-align`, `bool`
+* `--query-range.response-cache-ttl` `time.Duration`
+* `--query-range.response-cache-max-freshness` `time.Duration` default = `1m`
+* `--query-range.response-cache-config(-file)` `pathorcontent` + [CacheConfig](https://github.com/thanos-io/thanos/blob/55cb8ca38b3539381dc6a781e637df15c694e50a/pkg/store/cache/factory.go#L32)
+
+We plan to have in-mem, fifo and memcached support for now. Cache config will be exactly the same as the one used for Store Gateway.
+
+This command will be placeholder for any query planning or queueing logic that we might want to add at some point. It will be not part of any gRPC API.
+
+To make this happen we will propose a small refactor in Cortex code to avoid unnecessary package dependencies.
+
+### Alternatives
+
+#### Don't add anything, document Cortex query frontend and add examples of usage
+
+Unfortunately we tried this path already without success. Reasons were mentioned in [Motivation](202004_embedd_cortex_frontend.md#Motivation)
+
+#### Add response caching to Querier itself, in the same process.
+
+This will definitely simplify deployment if Querier would allow caching directly. However, this way is not really scalable.
+
+Furthermore, eventually frontend will be responsible for more than just caching. It is meant to do query planning like splitting or even
+advanced query parallelization (query sharding). This might mean future improvements in terms of query scheduling, queuing and retrying.
+This means that at some point we would need an ability to scale query part and caching/query planner totally separately.
+
+Last but not least splitting queries allows to perform request in parallel. Only if used in single binary we can achieve load balancing of those requests.
+
+NOTE: We can still consider just simple response caching inside the Querier if user will request so.
+
+#### Write response caching from scratch.
+
+I think this does not need to be explained. Response caching has proven to be not trivial. It's really amazing that we
+have opportunity to work towards something that works with experts in the field like @tomwilkie and others from Loki and Cortex Team.
+
+Overall, [Reusing is caring](https://www.bwplotka.dev/2020/how-to-became-oss-maintainer/#5-want-more-help-give-back-help-others).
+
+## Work Plan
+
+1. Refactor [IndexCacheConfig](https://github.com/thanos-io/thanos/blob/55cb8ca38b3539381dc6a781e637df15c694e50a/pkg/store/cache/factory.go#L32) to generic cache config so we can reuse.
+Make it implement Cortex cache.Cache interface.
+1. Add necessary changes to Cortex frontend
+    * Metric generalization (they are globals now).
+    * Avoid unnecessary dependencies.
+1. Add `thanos frontend` subcommand. Call it `query-frontend` in docs and in the communication.
```

Comment:
I thought the command will be `thanos query-frontend` i.e. the same name? :confused:

Reply:
Please read carefully the previous discussions: https://github.com/thanos-io/thanos/pull/2434#discussion_r409016044

Reply:
I was actually under the same impression

Reply:
Yea let's discuss. I think this what we did with `store-gateway`, so wanted to be consistent. Happy to change

Reply:
Talked offline, let's move to `query-frontend` subcommand as well. No harm, but only can help.